File: lib/loka/core/invariants.ex
defmodule Loka.Core.Invariants do
  @moduledoc """
  Pure checks for the invariants `protocol/invariants.json` assigns to `r3_pr5`, by id
  (docs/ROADMAP.md, verification harness). `kernel/ts/src/invariants.ts` is the TypeScript
  twin; both run the `"invariants"` cases of
  `protocol/fixtures/composition.json`.

  `check(id, observation)` is true when the invariant holds. Observation fields:

  - `"state"`, `"delta"`, `"result"`: a base state, a StateDelta and its
    `Loka.Core.Compose.compose/2` result;
  - `"resolution"`: a TargetResolution;
  - `"decision"`, `"commit"`, `"published"`: a DecisionResult, the host's commit outcome
    (`%{"status" => "committed", "revision" => n}`, `"failed"` or `"unknown"`) and what the
    host published. An unknown id raises.
  """
  alias Loka.Core.Compose

  @registry_path Path.expand("../../../protocol/error_registry.json", __DIR__)
  @external_resource @registry_path
  @registry File.read!(@registry_path)
  @evaluation_faults for e <- JSON.decode!(@registry),
                         e["category"] == "evaluation_fault",
                         do: e["code"]

  @spec check(String.t(), map()) :: boolean()
  def check("one_container_per_item", %{"state" => s, "result" => r}) do
    moved = moved(r)
    ids = Enum.map(moved, &elem(&1, 0))
    containers = Map.get(s, "containers", %{})

    ids == Enum.uniq(ids) and
      Enum.all?(moved, fn {e, c} -> is_binary(c) and Map.has_key?(containers, e) end)
  end

  def check("containment_acyclic", %{"state" => s, "result" => r}) do
    final = Map.merge(Map.get(s, "containers", %{}), Map.new(moved(r)))
    counts = Enum.frequencies(Map.values(final))

    Enum.all?(final, fn {e, c} -> not loops?(c, e, final, map_size(final)) end) and
      Enum.all?(Map.get(s, "capacities", %{}), fn {c, cap} -> Map.get(counts, c, 0) <= cap end)
  end

  def check("no_last_writer_wins", %{"delta" => %{"ops" => ops}, "result" => r}) do
    groups = Enum.group_by(ops, &Compose.key(Compose.target(&1)), & &1["writer_group"])
    Map.has_key?(r, "fault") or Enum.all?(groups, fn {_, gs} -> length(Enum.uniq(gs)) == 1 end)
  end

  # Checks only the read -> write value chain per target (fact value, container, quest state,
  # continuation or job status, clock, current resource value, cooldown start, barrier state), not capacity, revision, cycle or time bounds.
  def check("delta_preconditions_hold", %{"state" => s, "delta" => %{"ops" => ops}, "result" => r}) do
    Map.has_key?(r, "fault") or
      Enum.reduce_while(ops, %{}, fn op, seen ->
        k = Compose.key(Compose.target(op))
        {need, give} = link(op)
        before = Map.get_lazy(seen, k, fn -> initial(op, s) end)
        if before == need, do: {:cont, Map.put(seen, k, give)}, else: {:halt, false}
      end) != false
  end

  def check("fault_discards_whole_proposal", %{"result" => r}) do
    case r do
      %{"fault" => f} -> map_size(r) == 1 and Map.keys(f) -- ~w(kind code target) == []
      %{"changes" => _} -> map_size(r) == 1
    end
  end

  def check("fault_codes_are_evaluation_faults", %{"result" => r}) do
    case r do
      %{"fault" => %{"code" => code}} -> code in @evaluation_faults
      _ -> true
    end
  end

  def check("target_candidates_ordered", %{"resolution" => r}) do
    case r do
      %{"kind" => "ambiguous", "candidate_ids" => ids} ->
        ids |> Enum.chunk_every(2, 1, :discard) |> Enum.all?(fn [a, b] -> a < b end)

      _ ->
        true
    end
  end

  def check("no_proposed_event_escapes", %{"decision" => d, "commit" => c, "published" => p}) do
    allowed =
      case {d, c} do
        {%{"kind" => "accepted", "events" => es}, %{"status" => "committed", "revision" => n}} ->
          for e <- es, do: %{"committed_revision" => n, "event" => e}

        _ ->
          []
      end

    Enum.all?(p, &(&1 in allowed))
  end

  defp moved(r) do
    for %{"target" => %{"kind" => "containment", "entity_id" => e}, "value" => c} <-
          Map.get(r, "changes", []),
        do: {e, c}
  end

  # Walking up from e's container reaches e, or runs longer than there are rows (a cycle).
  defp loops?(nil, _, _, _), do: false
  defp loops?(e, e, _, _), do: true
  defp loops?(_, _, _, 0), do: true
  defp loops?(c, e, final, n), do: loops?(final[c], e, final, n - 1)

  # Each op reads one value of its target and leaves another: the fact value, the container,
  # the quest state, the continuation or job status, the clock.
  defp link(%{"op" => "fact.assign"} = op), do: {op["expected"], op["value"]}
  defp link(%{"op" => "entity.transfer"} = op), do: {op["source_id"], op["destination_id"]}
  defp link(%{"op" => "quest.activate"}), do: {nil, "active"}
  defp link(%{"op" => "quest.transition"} = op), do: {op["from"], op["to"]}
  defp link(%{"op" => "choice.open"}), do: {nil, "pending"}
  defp link(%{"op" => "choice.resolve"}), do: {"pending", "resolved"}
  defp link(%{"op" => "choice.close"}), do: {"pending", "closed"}
  defp link(%{"op" => "job.schedule"}), do: {nil, "pending"}
  defp link(%{"op" => "job.complete"}), do: {"pending", "completed"}
  defp link(%{"op" => "time.advance"} = op), do: {op["from"], op["to"]}
  defp link(%{"op" => "resource.adjust"} = op), do: {op["from"], op["to"]}
  defp link(%{"op" => "cooldown.start"} = op), do: {op["from"], op["at"]}
  defp link(%{"op" => "barrier.transition"} = op), do: {op["from"], op["to"]}

  defp initial(%{"op" => "fact.assign"} = op, s) do
    with nil <- get_in(s, ["facts", Compose.key(Compose.target(op))]),
         do: get_in(s, ["fact_defaults", Compose.key(op["fact"])])
  end

  defp initial(%{"op" => "entity.transfer", "entity_id" => e}, s),
    do: get_in(s, ["containers", e])

  defp initial(%{"op" => "quest." <> _, "instance_id" => i}, s),
    do: get_in(s, ["quests", i, "state"])

  defp initial(%{"op" => "choice." <> _, "continuation_id" => c}, s),
    do: get_in(s, ["choices", c, "status"])

  defp initial(%{"op" => "job." <> _, "job_id" => j}, s), do: get_in(s, ["jobs", j, "status"])
  defp initial(%{"op" => "time.advance"}, s), do: s["clock"]

  defp initial(%{"op" => "resource.adjust"} = op, s) do
    spec = get_in(s, ["resource_specs", Compose.key(op["resource"])])
    row = get_in(s, ["resources", Compose.key(Compose.target(op))])
    spec && Compose.current(row, spec, s["clock"])
  end

  defp initial(%{"op" => "cooldown.start"} = op, s),
    do: get_in(s, ["cooldowns", Compose.key(Compose.target(op))])

  defp initial(%{"op" => "barrier.transition"} = op, s) do
    with nil <- get_in(s, ["barriers", Compose.key(Compose.target(op))]),
         do: get_in(s, ["barrier_initial", Compose.key(op["barrier"])])
  end
end

File: kernel/ts/src/invariants.ts
// Pure invariant checks by id, twin of lib/loka/core/invariants.ex (its moduledoc states the
// observation fields). check(id, observation) is true when the invariant holds. The checks
// after STEP read one kernel step and are TypeScript only (rules are TypeScript, ADR-074).
import type { Json } from './canonical.ts';
import { current, key, same, target, type Result } from './compose.ts';
import {
  CAPABILITY_OWNERS,
  EVALUATION_FAULTS,
  type AdvertisedAction,
  type DeltaOp,
  type ExitView,
  type GameView,
} from './contracts.gen.ts';
import { validate } from './validate.ts';

// Observations are decoded JSON; fields are read loosely, as in the Elixir twin.
type Any = any;

const moved = (r: Result): [string, Json][] =>
  ('changes' in r ? r.changes : [])
    .filter((c) => c.target.kind === 'containment')
    .map((c) => [(c.target as Any).entity_id, c.value]);

// Walking up from e's container reaches e, or runs longer than there are rows (a cycle).
function loops(c: Json | undefined, e: string, final: Map<string, Json>): boolean {
  for (let n = final.size; n > 0; n--) {
    if (c === undefined) return false;
    if (c === e) return true;
    c = final.get(c as string);
  }
  return c !== undefined;
}

// Each op reads one value of its target and leaves another (the Elixir twin's link/1).
function link(op: Any): [Json | undefined, Json] {
  const fixed: Record<string, [Json | undefined, Json]> = {
    'quest.activate': [undefined, 'active'],
    'choice.open': [undefined, 'pending'],
    'choice.resolve': ['pending', 'resolved'],
    'choice.close': ['pending', 'closed'],
    'job.schedule': [undefined, 'pending'],
    'job.complete': ['pending', 'completed'],
  };
  if (fixed[op.op]) return fixed[op.op]!;
  if (op.op === 'fact.assign') return [op.expected, op.value];
  if (op.op === 'entity.transfer') return [op.source_id, op.destination_id];
  if (op.op === 'cooldown.start') return [op.from, op.at];
  return [op.from, op.to];
}

function initial(op: Any, s: Any): Json | undefined {
  const [family] = op.op.split('.');
  if (op.op === 'fact.assign') return s.facts?.[key(target(op))] ?? s.fact_defaults?.[key(op.fact)];
  if (op.op === 'entity.transfer') return s.containers?.[op.entity_id];
  if (family === 'quest') return s.quests?.[op.instance_id]?.state;
  if (family === 'choice') return s.choices?.[op.continuation_id]?.status;
  if (family === 'job') return s.jobs?.[op.job_id]?.status;
  if (family === 'cooldown') return s.cooldowns?.[key(target(op))];
  if (family === 'barrier')
    return s.barriers?.[key(target(op))] ?? s.barrier_initial?.[key(op.barrier)];
  if (family === 'resource') {
    const spec = s.resource_specs?.[key(op.resource)];
    return spec && current(s.resources?.[key(target(op))], spec, s.clock);
  }
  return s.clock;
}

const CHECKS: Record<string, (o: Any) => boolean> = {
  one_container_per_item: ({ state, result }) => {
    const m = moved(result);
    const ids = new Set(m.map(([e]) => e));
    const containers = state.containers ?? {};
    return (
      ids.size === m.length &&
      m.every(([e, c]) => typeof c === 'string' && Object.hasOwn(containers, e))
    );
  },
  containment_acyclic: ({ state, result }) => {
    const final = new Map<string, Json>([
      ...Object.entries<Json>(state.containers ?? {}),
      ...moved(result),
    ]);
    const counts = new Map<Json, number>();
    for (const c of final.values()) counts.set(c, (counts.get(c) ?? 0) + 1);
    return (
      [...final].every(([e, c]) => !loops(c, e, final)) &&
      Object.entries<number>(state.capacities ?? {}).every(
        ([c, cap]) => (counts.get(c) ?? 0) <= cap,
      )
    );
  },
  no_last_writer_wins: ({ delta, result }) => {
    const groups = new Map<string, Set<number>>();
    for (const op of delta.ops as DeltaOp[]) {
      const k = key(target(op));
      groups.set(k, (groups.get(k) ?? new Set()).add(op.writer_group));
    }
    return 'fault' in result || [...groups.values()].every((g) => g.size === 1);
  },
  // Checks only the read -> write value chain per target (a resource's current value, derived),
  // not capacity, revision, cycle, resource or time bounds.
  delta_preconditions_hold: ({ state, delta, result }) => {
    if ('fault' in result) return true;
    const seen = new Map<string, Json | undefined>();
    for (const op of delta.ops) {
      const k = key(target(op));
      const [need, give] = link(op);
      if (!same(seen.has(k) ? seen.get(k) : initial(op, state), need)) return false;
      seen.set(k, give);
    }
    return true;
  },
  fault_discards_whole_proposal: ({ result }) =>
    Object.keys(result).length === 1 &&
    (!('fault' in result) ||
      Object.keys(result.fault).every((k) => ['kind', 'code', 'target'].includes(k))),
  fault_codes_are_evaluation_faults: ({ result }) =>
    !('fault' in result) || EVALUATION_FAULTS.includes(result.fault.code),
  // EntityIds are ASCII by pattern, so string < is code-point order.
  target_candidates_ordered: ({ resolution }) =>
    resolution.kind !== 'ambiguous' ||
    resolution.candidate_ids.every(
      (id: string, i: number, ids: string[]) => i === 0 || ids[i - 1]! < id,
    ),
  no_proposed_event_escapes: ({ decision, commit, published }) => {
    const ok = decision.kind === 'accepted' && commit.status === 'committed';
    const allowed = ok
      ? decision.events.map((event: Json) => ({ committed_revision: commit.revision, event }))
      : [];
    return published.every((p: Json) => allowed.some((a: Json) => same(a, p)));
  },
  // STEP: `before` and `after` are the State around step(command) = decision; `view` is the
  // actor's GameView of the world before it.
  // A rejection or fault leaves the State (clock, containers, RNG, facts, resources, cooldowns,
  // barriers) byte for byte as it was (04 §5.0; §5.2 for a fault).
  rejection_consumes_nothing: ({ decision, before, after }) =>
    decision.kind === 'accepted' || same(before, after),
  // An unregistered command type is never accepted, and an accepted decision is valid against
  // its closed contract, so it names only registered delta ops, events and effects. Policy
  // operators are closed at the cartridge loader (lockStage).
  unknown_types_fail_closed: ({ command, decision }) =>
    decision.kind !== 'accepted' ||
    (Object.hasOwn(CAPABILITY_OWNERS.command, command.payload.type) &&
      validate('DecisionResult', decision).length === 0),
  // The view's entry for the command (its exit for a move, else its action) and admission agree
  // (04 §15, §19): available is never refused with a code the view shows for that entry;
  // unavailable is never accepted, and a refusal with such a code is the view's code.
  gameview_agrees_with_admission: ({ view, command, decision }) => {
    const entry = advertised(view, command.payload);
    const code = decision.kind === 'rejected' ? decision.error.code : undefined;
    const shown = SHOWN[command.payload.type] ?? [];
    if (!entry) return true;
    if (entry.available) return !shown.includes(code);
    return decision.kind !== 'accepted' && (!shown.includes(code) || code === entry.reason.code);
  },
};

// The codes the view can show on an entry: an exit's passage and fare; a recipe's policy and
// admission. ponytail: an engine verb's policy is always true, so it shows none; a cartridge
// action with a policy on an engine command joins when a cartridge authors one.
const SHOWN: Readonly<Record<string, readonly string[]>> = {
  move: ['exit_closed', 'exit_locked', 'insufficient_resource'],
  perform: ['invalid_state', 'cooldown', 'insufficient_resource'],
};

function advertised(view: GameView, p: Any): ExitView | AdvertisedAction | undefined {
  if (p.type === 'move') return view.exits.find((e) => e.direction === p.direction);
  const id = p.type === 'perform' ? undefined : (p.item_id ?? p.target_id);
  if (id === undefined) {
    const key = p.type === 'perform' ? p.action : p.type;
    return view.actions.find((a) => a.action_key === key);
  }
  const held = [...view.entities, ...view.inventory].find((e) => e.id === id);
  return held?.actions.find((a) => a.action_key === p.type);
}

/** True when the invariant holds; throws for an unknown id. */
export function check(id: string, observation: { [field: string]: unknown }): boolean {
  const f = CHECKS[id];
  if (!f) throw new Error(`unknown invariant ${id}`);
  return f(observation);
}
